### PR TITLE
Fix cluster replica unable to establish replication link in race condition

### PR DESCRIPTION
**Pull request by enjoy-binbin**
In d00b8af89265c501fb4a0cdd546702b90432a896, we fixed the bug that the
replica may set the primary host to '?'. Normally, we will correct '?'
to the correct IP when processing the next packet from primary, because
we will call nodeUpdateAddressIfNeeded when processing the packet.
```
int clusterProcessPacket(clusterLink *link) {
    ...
    /* Update the node address if it changed. */
    if (sender && type == CLUSTERMSG_TYPE_PING && !nodeInHandshake(sender) &&
        nodeUpdateAddressIfNeeded(sender, link, hdr)) {
        clusterDoBeforeSleep(CLUSTER_TODO_SAVE_CONFIG | CLUSTER_TODO_UPDATE_STATE);
    }
    ...
}
```

But there is a race will casue it to be unable to correct the primary host,
see the associated issue for more details about the race.

Simply put, if the replica just received a packet from another node that
contains new address about the primary, we will update primary node address
in here (the code i touched), when the replica receives the packet from
the primary, the check in nodeUpdateAddressIfNeeded will fail since the
address has been updated correctly, and we will not have the opportunity
to call replicationSetPrimary and update the primary host.

In places where the node address may be updated, we need to add this
replicationSetPrimary check to avoid this race.

### CODE REVIEW

**Review by zuiderkwast — APPROVED** [review]
Interesting. Was there a failing test case because of this?

The fix looks good to me, but I wish that Ping will take a look too.

Is there a risk that the gossip is wrong and that we should instead wait for a message from the primary?

**Comment by enjoy-binbin** [comment]
> Interesting. Was there a failing test case because of this?

no, i just meet a d00b8af89265c501fb4a0cdd546702b90432a896 issue and i looked at it in depth and think there is a race condition in here.

> Is there a risk that the gossip is wrong and that we should instead wait for a message from the primary?

you mean some node gossip a wrong address about the primary? Yes, it may happend, and then when the primary send its PING, the replica can update it via nodeUpdateAddressIfNeeded and everything work fine. The only downside i think is the replica will need to do a reconnect in here (a psync hopefully).

**Review by PingXie — APPROVED** [review]
This fix looks correct to me and is very reasonable for 8.0. 

however, conceptually, I feel that `nodeUpdateAddressIfNeeded` is the more appropriate abstraction that we need here and potentially for other similar cases.

**Comment by enjoy-binbin** [comment]
> however, conceptually, I feel that nodeUpdateAddressIfNeeded is the more appropriate abstraction that we need here and potentially for other similar cases.

you mean to call nodeUpdateAddressIfNeeded here, or find a way to merge them in one way? I think it's possible (i'll find time to look at it), the current code is OK to me now since it has comments and references, so people can easily see it. But a abstraction is also better i will try to if i can merge them

**Review by madolson — APPROVED** [review]
This seems necessary, it's whether or not there are other edge cases around here I'm less sure about. Each day makes me want to move towards a new cluster approach :/

**Comment by PingXie** [comment]
> you mean to call nodeUpdateAddressIfNeeded here, or find a way to merge them in one way?

Yes this is technically the "atomicity" contract that we need to enforce, i.e., any change to the node endpoint should be followed up by this check of `myself`'s primary.

That said, I think we should merge this one as-is but look into refactoring post 8.0.

>This seems necessary, it's whether or not there are other edge cases around here I'm less sure about. Each day makes me want to move towards a new cluster approach :/

I agree with you on a new cluster solution in many other cases but I think this is just a code bug :)

**Comment by wstar05** [comment]
Hi @enjoy-binbin,   I've noticed that the code you submitted in the current PR is highly similar to my contribution in https://github.com/redis/redis/pull/13572 (https://github.com/redis/redis/issues/13511), but my original author information has not been retained. Could you please add attribution or indicate the source? Thank you!

**Comment by enjoy-binbin** [comment]
? what do you mean? You means the source is your? Please note the date, i fix the issue even before you issue date. As far as i know, i started looking into this issue on 2024/08/28 and submitted the PR on 08/29

**Comment by wstar05** [comment]
> ? what do you mean? You means the source is your? Please note the date, i fix the issue even before you issue date. As far as i know, i started looking into this issue on 2024/08/28 and submitted the PR on 08/29

Sorry, I didn't notice the time. Please ignore it

**Comment by enjoy-binbin** [comment]
Please note that you should pick this in 6.2 first https://github.com/redis/redis/pull/10696
This PR fix another race i mentioned in the top comment
